Rejseplan api 2.0

Jawar19 opened the pull request:
## Breaking change
You will need to obtain an API key from labs.rejseplanen.dk in order to use the integration. This is a private license with 50.000 calls/month. 


## Proposed change
Complete update of backend library for the Rejseplanen integration as the service provider Rejseplanen.dk changed the API including an added authentication by API key. 


## Type of change
Updated backend to communicate with the new API provided by Rejseplanen.dk. The API is a customized HAFAS API for the Danish public transport system. The new backend is based on Multidepartureboard. Some old functionality is broken pending further updates to the backend library on pypi, see [pyRejsePlan](https://pypi.org/project/pyRejseplan/) and the [repo for pyRejsePlan](https://github.com/Jawar19/pyRejseplan)

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
There are no issues reported for the old API being deprecated by Rejseplanen.dk, but it have been discussed in Danish Home Assistant groups on Facebook. This is where I accepted the task.

This PR replaces the old pypi module completely as the maintainer of the old module could not be reached. Therefore there is no proper changelog between versions.
The new module pyRejsePlan implements API auth towards the API and datamodels using pydantic xml.

At the moment, i have tests within my own developed module testing my communication towards the API. I have not yet implemented automatic unit tests for the integration as these did not exist previously. Tests would also rely on an API key for updating snapshots. This would require me ask the developers at Rejseplanen if they would provide the project with a key for this purpose. 

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Jawar19 committed:
Update manifest.in with new codeowner and new deps

Jawar19 committed:
Update backend to new pydantic version of pyRejseplan

Jawar19 committed:
update to 0.1.17 backend

Jawar19 committed:
Merge remote-tracking branch 'upstream/dev' into Rejseplan_API_2_0

Jawar19 committed:
Merge branch 'home-assistant:dev' into Rejseplan_API_2_0

Jawar19 committed:
Merge remote-tracking branch 'upstream/dev' into Rejseplan_API_2_0

Jawar19 committed:
Merge remote-tracking branch 'upstream/dev' into Rejseplan_API_2_0

Jawar19 committed:
remove out commented code

Jawar19 committed:
Merge branch 'dev' into Rejseplan_API_2_0

joostlek reviewed (changes requested):
Oh that is so awesome to see. The new library also looks nice :)

But there are a few caveats and some things I want to write down first

1. First off, this is a legacy integration without config flow. We made the agreement not to change YAML config anymore. Which leads to a lot of integrations migrating to the config flow (which is IMO a great feature and a thing to have).
2. Something new that has been created in the recent times is sub entries. This would mean you can create 1 main entry with the API key and then add multiple lines or stops to that entry as a sub entry. this way you don't have to create a new entry for every line or stop. Whenever your API would go invalid, you can just enter it once instead of having to go through 10 reauth flows. I think transport integrations are a really good fit for this feature.
3. What can we reuse from the current data? Because I see that you now remove `CONF_ROUTE` and `CONF_DIRECTION`. When we migrate to a config flow, ideally we import the existing data and make it easy for the user.
4. For how long is the old intergation not working? Because if this is a long long long time that might open up other doors.

And please don't be daunted by the fact that this makes the task bigger, I think this is a doable task and we're here to support you :)

joostlek commented:
> My plan was initially get the integration up and running, while keeping the PR as small as possible as per the PR dev docs. Then make another PR with config flow.

Right, but that is a bit difficult as we promised to the community to not break existing YAML configs and this would violate that promise (and before you ask, reusing an old field also counts as breaking in a way).

> Would I be able to make one update call for all subentries and then filter them for each subentry? I still have a lot to learn about HA.

I think that is a possibility yes, I am not 100% sure about how this integration itself works, but this doesn't sound like a strange use case.

> The old integration/API have been down for a year. I have been busy and i couldn't get any other Danish dev to join me in this endeavor.

Oh but that is fine, I think this gives us a bit more room to breathe as we can assume that there are very little people who still have this in their YAML and thus rely on this. However this doesn't mean that we should disregard the importing of old configuration (as that is also one of the promises we made to the community)


I know the dutch railway integration is also in the process of getting this exact same migration going on (except that API still works). So I think I should invest a bit more time into that one so you also have a better example of what to expect and what a proper direction is, since subentries are quite new and I don't have nice examples

joostlek commented:
Well, we need to import the existing configuration into a config entry. My current assumption is that it's likely easier to implement sub entries right away as it will save you some migrations

Jawar19 committed:
Merge remote-tracking branch 'upstream/dev' into Rejseplan_API_2_0

Jawar19 committed:
First attempt on implementing subentries

during process several errors in datahandling from the API was fixed during debugging of the subflows. Backend have been updated as well.

Jawar19 committed:
Fixed bad string in strings.json

Jawar19 committed:
fix subentries config flow

Jawar19 committed:
Fix user button description in sub flow

Jawar19 committed:
add better description for subentry

Jawar19 committed:
Add reload on creation of new stop

Jawar19 committed:
Add correct callback for update from coordinator

Jawar19 committed:
Changed log message to debug level

Jawar19 committed:
Add minute tick update callback for sensor

In order for the sensor to update every minute to do proper countdown, a timer have been added firing an event every 1 minute to update state.

Jawar19 committed:
update pyrejseplane to 0.1.22

Jawar19 committed:
Update backend to 0.1.23

Jawar19 committed:
Change default name to use CONSTANT

Jawar19 committed:
added direction entry type to complete current working attributes.

Route does not yet work as the backend does not support it yet.

Jawar19 committed:
Add test for config flow and subentries

Jawar19 committed:
Merge remote-tracking branch 'upstream/dev' into Rejseplan_API_2_0

Jawar19 committed:
Merge branch 'dev' into Rejseplan_API_2_0
